### PR TITLE
feat: add drizzle

### PR DESCRIPTION
**Pull request by juliusmarminge**
Closes #<issue>

## ✅ Checklist

- [ ] I have followed every step in the [contributing guide](https://github.com/t3-oss/create-t3-app/blob/main/CONTRIBUTING.md) (updated 2022-10-06).
- [ ] The PR title follows the convention we established [conventional-commit](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] I performed a functional test on my final commit

---

## Changelog

_[Short description of what has changed]_

---

## Screenshots

_[Screenshots]_

💯

### CODE REVIEW

File: cli/template/extras/config/drizzle.config.ts
Comment (juliusmarminge) [inline cli/template/extras/config/drizzle.config.ts:1]:
```suggestion
import { type Config } from "drizzle-kit";
```

File: cli/src/installers/dependencyVersionMap.ts
Comment (c-ehrlich) [inline cli/src/installers/dependencyVersionMap.ts:18]: not sure how i feel about this considering the opinion we've had in the past that the "core stack" should always be free. 

is the reasoning here that this is our best way to be edge ready?

File: cli/template/extras/src/server/api/routers/example/with-auth-drizzle.ts
Comment (c-ehrlich) [inline cli/template/extras/src/server/api/routers/example/with-auth-drizzle.ts:18]: is this correct?

File: cli/template/extras/src/server/api/routers/example/with-drizzle.ts
Comment (c-ehrlich) [inline cli/template/extras/src/server/api/routers/example/with-drizzle.ts:13]: is this correct?

**Review by c-ehrlich — COMMENTED** [review]
will do a functional test tomorrow :)

File: cli/src/cli/index.ts
Comment (juliusmarminge) [reply cli/src/cli/index.ts:169]: this is just for the CI flags

when you're using the CLI it will keep prompting you if you add invalid stuff (see `validate` in the `promptPackages`:

image

File: cli/src/helpers/logNextSteps.ts
Comment (juliusmarminge) [reply cli/src/helpers/logNextSteps.ts:31]: yes

File: cli/src/installers/dependencyVersionMap.ts
Comment (juliusmarminge) [reply cli/src/installers/dependencyVersionMap.ts:18]: idk we could take it one step further and prompt for what driver they want - but that's though when it comes to the nextauth adapter as it will differ a bit.

until nextauth ships the adapter and handles all of this for you, maybe we can just go for pscale?

the biggest selling point here is the edge compat and selecting the normal mysql driver wouldn't make sense for that.

idk what's best - hence i left the comment pointing it out - but definetely something to think about

File: cli/template/extras/src/server/api/routers/example/with-auth-drizzle.ts
Comment (juliusmarminge) [reply cli/template/extras/src/server/api/routers/example/with-auth-drizzle.ts:18]: yep 😎 

CleanShot 2023-06-04 at 21 33 54@2x

File: cli/template/extras/src/server/api/routers/example/with-drizzle.ts
Comment (juliusmarminge) [reply cli/template/extras/src/server/api/routers/example/with-drizzle.ts:13]: ^^

**Review by t3dotgg — COMMENTED** [review]
next-auth + drizzle isn't a super happy story at the moment 😓

**Comment by juliusmarminge** [comment]
> next-auth + drizzle isn't a super happy story at the moment 😓 

Why? Worked when I tried it 🙄

File: .changeset/fast-teachers-explain.md
Comment (michalbundyra) [inline .changeset/fast-teachers-explain.md:18]:
```suggestion
     opts.ctx.db.query.example.findMany()
```

**Comment by FleetAdmiralJakob** [comment]
Fixes #1452

File: cli/template/extras/config/drizzle.config.ts
Comment (juliusmarminge) [reply cli/template/extras/config/drizzle.config.ts:8]: Drizzle doesnt load the envs themselves (this config is used for drizzle-kit i.e when you db:push)

**Review by c-ehrlich — CHANGES_REQUESTED** [review]
For me this requires one more thing to be ready to release: a dev db for drizzle.

some thoughts/questions related to this:
- drizzle requires much bigger differences in how the schema are defined between different dialects of sql. see their [mysql](https://orm.drizzle.team/docs/column-types/mysql) vs [sqlite](https://orm.drizzle.team/docs/column-types/sqlite) docs.
- this makes our current solution of shipping a sqlite database and then letting people switch to whatever they want less ideal - imo we are being pushed into prescribing a default sql dialect which is the same for dev and prod.
- i believe this dialect should be mysql - it's a database that we all like, and it's the database planetscale uses, which is already very relevant now and might become even more relevant with edge + their adapter.
- we could pivot to something like turso which would enable sqlite everywhere, but although turso seems awesome i'm personally leaning against this
- no matter what our db/orm is, we should always provide a working dev db setup. this setup should work offline, ie "just provision a planetscale instance for dev" is not a valid solution.

where this seems to lead me is a mysql docker container that is started alongside the dev server. i don't love this, as developing completely without docker feels great. but i don't see any other compelling options for drizzle.

thoughts? any alternatives that i'm forgetting? also if we do this, should we also do the same for prisma?

(there's other things that could probably be improved as well, but i'm happy to leave them for future PRs so we can get this out, and merged into the app dir PR)

File: cli/src/cli/index.ts
Comment (GermanJablo) [reply cli/src/cli/index.ts:169]: Wouldn't a separate question be better? If the default settings are what most users use, it's comfortable to move forward without thinking about the options.

**Comment by matibox** [comment]
nextauthjs/next-auth#8258 https://authjs.dev/reference/adapter/drizzle

Next Auth now has an official drizzle adapter

**Comment by juliusmarminge** [comment]
This comment is buried way up but here's the latest one for y'all:

```sh
pnpm create t3-app@7.18.1-beta.19d5f83
```

File: cli/src/cli/index.ts
Comment (c-ehrlich) [inline cli/src/cli/index.ts:6]: two commented imports here

File: cli/src/installers/nextAuth.ts
Comment (c-ehrlich) [inline cli/src/installers/nextAuth.ts:15]: isnt it out now?

File: cli/template/extras/src/server/db/drizzle-schema-auth.ts
Comment (c-ehrlich) [inline cli/template/extras/src/server/db/drizzle-schema-auth.ts:20]: could we be slick and replace `project1` with the name of the inited project?

File: cli/src/cli/index.ts
Comment (c-ehrlich) [reply cli/src/cli/index.ts:246]: how much additional work would it be? maintaining separate `index-drizzle.ts`, `drizzle-schema-base.ts`, and `drizzle-schema-auth.ts` for each dialect doesn't seem like too big of a deal.

here's roughly the list of possible things that feel like they could possibly be included (doesn't mean we should include all of them)

- dev: planetscale, prod: planetscale (what we already have)
- dev: mysql container, prod: mysql connection url
- dev: mysql container, prod: planetscale (is this even a good idea?)
- dev: postgres container, prod: postgres connection url
- (we could also consider setups with other serverless/edge ready databases)

some of these require docker setups, which means at least a `docker-compose.yml`, and maybe also changes to `package.json` to create a smooth dev experience.

i'm not for or against any of this, not sure what's the right way to go. what do you think?

File: cli/src/cli/index.ts
Comment (c-ehrlich) [reply cli/src/cli/index.ts:174]: i think i'm in favour of exiting with a non-zero code. 2 (Invalid command line usage or invalid input) seems to make sense.

why do you think we shouldn't? does it relate to our own actions that make sure the CLI works in CI?

Comment (juliusmarminge) [reply cli/src/cli/index.ts:174]: > why do you think we shouldn't? does it relate to our own actions that make sure the CLI works in CI?

Yes, it would probably be more work to make the CI matrix work in that case, maybe we can have a check in a catch block or something that just gracefully exits if `process.env.CI` is true?

File: cli/src/cli/index.ts
Comment (juliusmarminge) [reply cli/src/cli/index.ts:246]: I think either way it can be a follow-up in another release. Dont need to block this one before we have made the final decision?

Comment (c-ehrlich) [reply cli/src/cli/index.ts:246]: agreed

File: cli/src/cli/index.ts
Comment (c-ehrlich) [reply cli/src/cli/index.ts:174]: yea sounds good

**Comment by juliusmarminge** [comment]
Looks like we can't quite do that @c-ehrlich.

Can we somehow detect this exit in the action and exit the action gracefully from there?

Or maybe we can add an if-statement to the action
```yaml
- name: Check valid config
  if: ${{ matrix.prisma == "true" && matrix.drizzle == "true" }} # unsure of exact syntax
  run: # graceful exit
```

**Comment by juliusmarminge** [comment]
Nice, now the Drizzle + Prisma exits early: https://github.com/t3-oss/create-t3-app/actions/runs/6052746581/job/16426820256?pr=1461

**Review by c-ehrlich — APPROVED** [review]
lookin good

File: www/src/pages/pl/faq.md
Comment (AndriiSherman) [inline www/src/pages/pl/faq.md:21]: If create-t3-app is using drizzle-orm itself, I think it would be better to have a link to drizzle-orm docs rather than the drizzle team. I guess this one should be much better: https://orm.drizzle.team/docs/overview. We have also updated an overview page and the documentation itself.

File: www/src/pages/en/faq.mdx
Comment (AndriiSherman) [inline www/src/pages/en/faq.mdx:24]: If create-t3-app is using drizzle-orm itself, I think it would be better to have a link to drizzle-orm docs rather than the drizzle team. I guess this one should be much better: https://orm.drizzle.team/docs/overview. We have also updated an overview page and the documentation itself.

File: cli/template/extras/src/server/db/drizzle-schema-base.ts
Comment (AndriiSherman) [inline cli/template/extras/src/server/db/drizzle-schema-base.ts:24]: We will change the documentation examples for MySQL to use `bigint().primaryKey().autoincrement()` instead of serial. Using SERIAL may not explicitly convey the use of a bigint autoincrement field and could lead to issues. We will also make the same adjustment for PostgreSQL, especially since PostgreSQL is beginning to recommend IDENTITY columns over SERIAL column types. While we won't deprecate these options, we don't want to feature them in our examples or promote them as the recommended way to set up a schema

So it would be great, if create-t3-app will also recommend this id columns:
```ts
id: bigint("id", { mode: "number" }).primaryKey().autoincrement()
```
We will update drizzle orm docs with this recommendation as well

File: cli/template/extras/src/server/db/drizzle-schema-auth.ts
Comment (AndriiSherman) [inline cli/template/extras/src/server/db/drizzle-schema-auth.ts:26]: We will change the documentation examples for MySQL to use bigint().primaryKey().autoincrement() instead of serial. Using SERIAL may not explicitly convey the use of a bigint autoincrement field and could lead to issues. We will also make the same adjustment for PostgreSQL, especially since PostgreSQL is beginning to recommend IDENTITY columns over SERIAL column types. While we won't deprecate these options, we don't want to feature them in our examples or promote them as the recommended way to set up a schema

So it would be great, if create-t3-app will also recommend this id columns:
```ts
id: bigint("id", { mode: "number" }).primaryKey().autoincrement()
```
We will update drizzle orm docs with this recommendation as well

**Review by AndriiSherman — COMMENTED** [review]
I'd like to share some ideas for useful changes to Drizzle in `create-t3-app`

File: www/src/pages/pl/faq.md
Comment (juliusmarminge) [reply www/src/pages/pl/faq.md:21]: Awesome, we'll add proper docs later so we can adress that then I suppose

File: cli/template/extras/src/server/db/drizzle-schema-base.ts
Comment (juliusmarminge) [reply cli/template/extras/src/server/db/drizzle-schema-base.ts]: I changed it to use bigint, would love for unsigned support in the near future though 🙏🏼

Comment (AndriiSherman) [reply cli/template/extras/src/server/db/drizzle-schema-base.ts]: These recommendations are intended for new users, which is why we will update our documentation for now (while there is no unsigned option). Furthermore, the unsigned option should be included in the next release, so in any case, it will be available soon.

I'm curious about why you want to have the unsigned option and why bigint alone is not sufficient. BIGINT has a range of up to 9,223,372,036,854,775,807. I understand that with unsigned, you can extend this range by a factor of two, but even with a signed option, I can imagine that there may be so many rows in a table that reaching 9,223,372,036,854,775,807 would be challenging.

Consider, for example, a messages table receiving 10 billion messages per day. In this scenario, it would take approximately 2,479,401 years to reach the limits. This calculation assumes we are discussing changing from serial to bigint

Comment (juliusmarminge) [reply cli/template/extras/src/server/db/drizzle-schema-base.ts]: Yea obviously a BIGINT will be enough for most apps, but I just like being explicit what data will be in the column, and since we won't have any negatives it just makes sense that it's unsigned...

Comment (AndriiSherman) [reply cli/template/extras/src/server/db/drizzle-schema-base.ts]: Agree! 
Will add support for it soon!

Comment (juliusmarminge) [reply cli/template/extras/src/server/db/drizzle-schema-base.ts]: I think we're good to merge this then!

Exciting, app dir next :)

**Comment by juliusmarminge** [comment]
Ship 🚢

**Comment by checkerschaf** [comment]
Thanks a lot to all of you who have contributed to this 🙏

**Pull request merged**
